core/src/main/scala/cats/derived/traverse.scala

```diff
@@ -1,166 +1,132 @@
 package cats.derived
 
-import cats.syntax.all._
-import cats.{Applicative, Eval, Now, Traverse}
+import cats.{Applicative, Eval, Traverse}
 import shapeless._
 
 import scala.annotation.implicitNotFound
 
-/**
-  * Based on the `MkFoldable` implementation.
-  */
 @implicitNotFound("Could not derive an instance of Traverse[${F}]")
 trait MkTraverse[F[_]] extends Traverse[F] {
-
-  def safeTraverse[G[_] : Applicative, A, B](fa: F[A])(f: A => Eval[G[B]]): Eval[G[F[B]]]
-
+  def safeTraverse[G[_]: Applicative, A, B](fa: F[A])(f: A => Eval[G[B]]): Eval[G[F[B]]]
   def safeFoldLeft[A, B](fa: F[A], b: B)(f: (B, A) => Eval[B]): Eval[B]
-
   def foldRight[A, B](fa: F[A], lb: Eval[B])(f: (A, Eval[B]) => Eval[B]): Eval[B]
 
-  def traverse[G[_] : Applicative, A, B](fa: F[A])(f: A => G[B]): G[F[B]] =
-    safeTraverse(fa)(a => Now(f(a))).value
+  def traverse[G[_]: Applicative, A, B](fa: F[A])(f: A => G[B]): G[F[B]] =
+    safeTraverse(fa)(a => Eval.later(f(a))).value
 
   def foldLeft[A, B](fa: F[A], b: B)(f: (B, A) => B): B =
-    safeFoldLeft(fa, b) { (b, a) => Now(f(b, a)) }.value
+    safeFoldLeft(fa, b)((b, a) => Eval.later(f(b, a))).value
 
+  // The default `forall` is not lazy.
+  override def forall[A](fa: F[A])(p: A => Boolean): Boolean =
+    foldRight(fa, Eval.True)((a, lb) => lb.map(_ && p(a))).value
```

Comment:
interesting. Did you find out because it broke the `forallLazy` law [here](https://github.com/typelevel/cats/blob/master/laws/src/main/scala/cats/laws/UnorderedFoldableLaws.scala#L33)?

Reply:
Update, yes I confirmed that without these it broke the law. However the default `forall` implementation did not break this law on List instance (yes, I commented out the override forall in the list instance). I am trying understand is the default implementation of `forall` incorrect? or is it the foldRight derived here different?

Reply:
I figured it out. Our foldRight seems to be problematic (i.e. we shouldn't override the forall and exist, we should fix our `foldRight`)
I added this test to our `TraverseSuite` on your branch
```scala
  test(s"$context foldRightLazy Not Empty") {
      var i = 0
      iList.foldRight(large, Eval.now(0)) { (_, _) =>
        i += 1
        Eval.now(1)
      }.value

      assert(i == 1)
    }
```
This test failed , `i` turns out to be 9999 the size of large. 
It is a surprise that it's caught through foralllazy and existlazy laws, which I think is an issue in cats-laws. So I PRed one there (https://github.com/typelevel/cats/pull/2817)

I haven't got the time to figure which part in our derivation that makes foldRight not lazy. Need to go afk for the night.

Reply:
Hmm, interesting. I don't understand entirely how it's supposed to work. How do you know that the first value is the final one so you stop calling foldRight?

Reply:
I figured it out. `foldRight` was being called in the wrong order for HCons. I noticed this before but I missed the part where I have to add `Eval.defer` to make it lazy / stack safe:

https://github.com/typelevel/kittens/compare/8223a1c8a9ad16b64d0cf6b3ff1b3ef9527db579..1f85bcff82661e3476b0ac361bb19f88bedb9ef6